src/FSharpPlus/Extensions/Map.fs

```diff
@@ -57,11 +57,9 @@ module Map =
     let mapValues3 f (x: Map<'Key, 'T1>) (y: Map<'Key, 'T2>) (z: Map<'Key, 'T3>) = Map <| seq {
         let f = OptimizedClosures.FSharpFunc<_,_,_,_>.Adapt f
         for KeyValue(k, vx) in x do
-            match Map.tryFind k y with
-            |Some vy -> match Map.tryFind k z with
-                        | Some vz -> yield (k, f.Invoke (vx, vy, vz))
-                        | None -> ()
-            | None -> ()}
+            match Map.tryFind k y, Map.tryFind k z with
+            | Some vy, Some vz -> yield (k, f.Invoke (vx, vy, vz))
+            | _      , _       -> () }
```

Comment:
This reads much better but the second `tryFInd` will execute unnecessarily in some cases.

Here's a possible solution:
 
```
            match Map.tryFind k y, lazy (Map.tryFind k z) with
            | Some vy, Lazy (Some vz) -> yield (k, f.Invoke (vx, vy, vz))
            | _      , _              -> () }
```